src/server/input.ts

```diff
@@ -50,14 +51,98 @@ export class Keyboard {
     this._page = page;
   }
 
+  async _updateKeyboardIndicator(description?: KeyDescription) {
+    if (!this._page._browserContext._browser._options.debugKeyboard)
+      return;
+    let text;
+    let dim = true;
+    if (!description) {
+      if (this._currentText)
+        return;
+      this._currentText = '';
+      text = '';
+    } else if (!this._pressedModifiers.size && description.key.trim().length === 1) {
+      this._currentText += description.key;
+      text = this._currentText;
+      dim = false;
+    } else {
+      text = description.code;
+      this._currentText = '';
+    }
+    if (text)
+      text = [...this._pressedModifiers, text].join(' + ');
+    await this._renderKeyboardIndicator(text, dim);
+  }
+
+  async _renderKeyboardIndicator(text: string, dim = false) {
+    (await this._page.mainFrame()._utilityContext()).evaluateInternal(({text, dim}) => {
+      const element: HTMLElement = document.querySelector(':scope > playwright-keyboard-indicator') || createKeyboardIndicator();
+      (element as any).__setText(text, dim);
+
+      function createKeyboardIndicator() {
+        const element = document.createElement('playwright-keyboard-indicator');
+        const shadow = element.attachShadow({mode: 'closed'});
+        shadow.innerHTML = `
```

Comment:
our injected script already has html library for debug overlays, let's align those implementations.